SSA/Experimental/Bits/Fast/README.md

```diff
@@ -0,0 +1,94 @@
+# Automata Tactic
+
+This is documentation for how the automata tactic works. This tactic deserves documentation because it is non-trivial and contains some difficult concepts.
+
+By analogy, the automata tactic is like Presberger Arithmetic. Just like Presberger Arithmetic only involves constant-memory operations on natural numbers,
+the Automata Tactic can automatically solve expressions on BitVectors which involve constant-memory operations.
+
+# Goal of the Automata Tactic
+
+The automata tactic is a decision procedure. It takes in statments of the form
+
+
+\[
+\forall (w : \mathbb{N})
+\forall (x_1, x_2, x_3, \ldots x_n : BitVec w)
+(x_1 &&& x_2) + (x_1 ||| x_2) = (x_1 + x_2)
+\]
+
+The automata tactic only accepts expressions containing finite memory operations, or operations which can 
+be implmeneted with an amount of extra memory scratch space which does not depend on the size of the inputs.
+For example, the automata tactic works with
+
+1) Bitwise operations (which require 0 extra bits of memory scratch space)
+2) Addition and subtraction (which require 1 extra bit of memory scratch space)
+3) negation (which requires 1 extra bit of memory scratch space)
+
+But crucially, the automata tactic does not support multiplication, because the amount of memory
+required to do a multiplication depends on the size of the inputs.
+
+The reason that this tactic only accepts operations of constant memory is that this corresponds to state machines with a finite number of states.
+If we accepted multiplication as an operation, the number of states in the state machine could potentially blow up.
+
+# How is the Automata Tactic Implemented?
+
+To figure out if 
+
+\[a = b\]
+
+we ask:
+Does the automaton 
+
+\[automaton(a ^^^ b)\]
+
+accept the string of all zeros?
```

Comment:
this seems to contradict with the definition below involving arities > 0. I would guess that the thing is nondeterministic over the free variable transitions, forming an NFA, and then we ask if there is any non-deterministic path where the NFA rejects. (opposite to the usual notion of NFAs, where they accept iff there is some path to the accept state, here we are actually asking whether there is any path to the reject state.) I'm guessing you then convert the NFA to a DFA and you have to check whether the resulting DFA accepts iff the input is all zeros of some length.